SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Storage/Storage.cs

```diff
@@ -369,18 +367,40 @@ private Task<T> AddTaskToQueue<T>(Func<T> action)
                 {
                     return action();
                 }
-                catch (Exception e) when (!(e is StorageException))
+                catch (Exception e)
                 {
-                    // Tasks should already be throwing only storage exceptions, but in case any are missed, 
-                    // which has happened, catch them here and wrap in a storage exception. This will prevent 
-                    // the exception from being unobserved.
-                    throw new StorageException(e);
+                    // And regarding the comment in other variant of this function, async lambda + await cannot work with a Func anyway.
+                    throw HandleStorageRelatedExceptionAsync(e).GetAwaiter().GetResult();
                 }
             });
             AddTaskToQueue(task);
             return task;
         }
 
+        private async Task<Exception> HandleStorageRelatedExceptionAsync(Exception e)
+        {
+            // Check if database is corrupted, we have evidence (https://github.com/microsoft/appcenter-sdk-dotnet/issues/1184)
+            // that it's not always originated by a proper SQLiteException (which would then be converted to StorageException in StorageAdapter).
+            // If it was always the right type then the exception would not have been unobserved in that application before we changed the re-throw logic here.
+            // But the message is definitely "Corrupt" and thus unfortunately that is the only check we seem to be able to do as opposed to type/property checking.
+            if (e.Message == "Corrupt")
+            {
+                AppCenterLog.Error(AppCenterLog.LogTag, "Database corruption detected, deleting the file and starting fresh...", e);
+                await _storageAdapter.DeleteDatabaseFileAsync();
+                await InitializeDatabaseAsync();
+            }
+
+            // Return exception to re-throw.
+            if (!(e is StorageException))
```

Comment:
Tried to find how to simplify it (drop `()`) and found [this](https://stackoverflow.com/questions/529944/check-if-object-is-not-of-type-equivalent-for-is-c-sharp#comment67093273_529959)
Interesting variant, not sure what's better. 
BTW Kotlin has [a nice construction for this case](https://kotlinlang.org/docs/reference/typecasts.html#is-and-is-operators) - `!is`

```suggestion
            if (false == e is StorageException)
```

Reply:
I don't see this as an improvement.

Reply:
Reversing logic sounds better, will push that instead.